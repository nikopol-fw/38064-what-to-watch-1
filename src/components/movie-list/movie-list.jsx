// movie-list.jsx

import React from 'react';
import PropTypes from 'prop-types';

import {MovieCard} from '../movie-card/movie-card';


export class MovieList extends React.PureComponent {
  constructor(props) {
    super(props);
    this.activationDuration = 1000;

    this.state = {
      activeCard: null,
      timer: null,
    };
  }

  render() {
    const {movies} = this.props;

    return <div className="catalog__movies-list">
      {movies.map((item, i) => <MovieCard key={i}
        title={item.title}
        src={item.src}
        preview={item.preview}
        isActive={i === this.state.activeCard ? true : false}
        onMouseEnter={() => {
          this._setActiveCard(i);
        }}
        onMouseLeave={() => {
          this._unsetActiveCard();
        }}
      />)}
    </div>;
  }

  _setActiveCard(card) {
    const timer = setTimeout(() => {
      this.setState({
        activeCard: card,
      });
    }, this.activationDuration);

    this.setState({timer});
  }

  _unsetActiveCard() {
    if (this.state.timer) {
      clearTimeout(this.state.timer);
    }

    this.setState({
      activeCard: null,
      timer: null,
    });
  }
}


MovieList.propTypes = {
  movies: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string,
    src: PropTypes.string,
    preview: PropTypes.string,
  }))
};
