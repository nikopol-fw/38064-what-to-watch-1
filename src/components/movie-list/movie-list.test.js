// movie-list.test.jsx

import React from 'react';
import renderer from 'react-test-renderer';

import {MovieList} from './movie-list';


const mock = [
  {
    title: `Fantastic Beasts: The Crimes of Grindelwald`,
    src: `fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }, {
    title: `Bohemian Rhapsody`,
    src: `bohemian-rhapsody.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  }, {
    title: `Macbeth`,
    src: `macbeth.jpg`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }, {
    title: `Aviator`,
    src: `aviator.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  }, {
    title: `We need to talk about Kevin`,
    src: `we-need-to-talk-about-kevin.jpg`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }
];


it(`MovieList correctly renders`, () => {
  const tree = renderer
    .create(<MovieList
      movies={mock}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
