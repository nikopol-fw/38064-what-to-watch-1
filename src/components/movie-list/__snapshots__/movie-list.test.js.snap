// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList correctly renders 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        autoPlay={false}
        height={175}
        loop={true}
        muted={true}
        poster="/img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
        preload="metadata"
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
        width={280}
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Fantastic Beasts: The Crimes of Grindelwald
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        autoPlay={false}
        height={175}
        loop={true}
        muted={true}
        poster="/img/bohemian-rhapsody.jpg"
        preload="metadata"
        src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
        width={280}
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Bohemian Rhapsody
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        autoPlay={false}
        height={175}
        loop={true}
        muted={true}
        poster="/img/macbeth.jpg"
        preload="metadata"
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
        width={280}
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Macbeth
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        autoPlay={false}
        height={175}
        loop={true}
        muted={true}
        poster="/img/aviator.jpg"
        preload="metadata"
        src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
        width={280}
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Aviator
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        autoPlay={false}
        height={175}
        loop={true}
        muted={true}
        poster="/img/we-need-to-talk-about-kevin.jpg"
        preload="metadata"
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
        width={280}
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        We need to talk about Kevin
      </a>
    </h3>
  </article>
</div>
`;
